ngraph_bridge/enable_variable_ops/ngraph_rewrite_for_tracking.cc

```diff
@@ -71,34 +69,75 @@ Status RewriteForTracking(Graph* graph, int graph_id) {
             IsRefType(edge->dst()->input_type(edge->dst_input()))) {
           just_looking = false;
           // if the output reference is read by NGraph supported ops, do not
-          // turn off is_tf_just_looking
+          // add the sync node
           if (!IsNGVariableType(edge->dst()->type_string())) {
-            NGRAPH_VLOG(1)
-                << DebugNode(edge->dst())
-                << "needs reference, setting is_tf_just_looking to false";
-            is_tf_just_looking = false;
+            NGRAPH_VLOG(1) << DebugNode(edge->dst())
+                           << "needs reference from ( "
+                           << DebugNode(edge->src())
+                           << " ), adding a "
+                              "NGraphVariableUpdateNGTensor node here";
+
+            // If the number of outputs for the TF optimizer > 1
+            // we do not handle it and error out
+            if (node->num_outputs() > 1) {
+              return errors::InvalidArgument(
+                  "The TF optimizer has more than 1 output ",
+                  DebugNode(edge->dst()));
+            }
+
+            // Since the dst node takes in this variable as a reference
+            // and is not supported by NGraph, it might update the
+            // variable hence the sync node is required here.
+            NodeBuilder::NodeOut input_ref;
+            input_ref = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+            DataType dtype;
+            TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "dtype", &dtype));
+            string shared_name;
+            TF_RETURN_IF_ERROR(
+                GetNodeAttr(node->attrs(), "shared_name", &shared_name));
+            string sync_node_name = node->name() + "_sync_node";
+            Node* sync_node;
+            NodeBuilder nb =
+                NodeBuilder(sync_node_name, "NGraphVariableUpdateNGTensor")
+                    .Input(input_ref)
+                    .Attr("ngraph_graph_id", graph_id)
+                    .Attr("ngraph_variable_shared_name", shared_name)
+                    .Attr("T", dtype)
+                    .Device(node->assigned_device_name());
+            Status status = nb.Finalize(graph, &sync_node);
+            TF_RETURN_IF_ERROR(status);
+            sync_node->set_assigned_device_name(node->assigned_device_name());
+
+            // Connect output edges from the TF optimizer to the sync node
+            // which will be for the input index 0 since that is the
+            // output we want.
+            // This should not replace the control edges.
+            if (edge->dst_input() == 0) {
+              TF_RETURN_IF_ERROR(
+                  ReplaceOnlyDataOutputEdges(graph, edge->dst(), sync_node));
```

Comment:
I would think all the ctrl + data output edges outgoing of optimizer node need to be moved to the sync_node